### DISCUSSION TITLE
User-defined null on structs

### DISCUSSION THREAD

**Discussion by Perksey** [dotnet/csharplang#7602] [Language Ideas]
# Summary
In the advent of ref structs, there are now many use cases for which the developer may want to define what "null" means for their struct. For example, users today can write `null` in place of an expression that evaluates to `ReadOnlySpan<T>` but this is only because of an implicit operator converting arrays to spans, and converting a null array to a span yields an empty, nullref span. It is likely that the majority of users do not know that this is how `null` on spans work, and therefore only care for the presence of `null` for these types of structs.

There may be use cases where user types are unable to wield `null` as-is as there may be multiple implicit operators that `null` could be applicable for.

This suggestion is for adding a way for users to control what occurs when a developer writes `null` for their structs. 

Spit-balling here:
```cs
public ref struct Maybe
{
    public static implicit operator null() => new Maybe(/* ... */);
}
public ref struct Maybe
{
    public static implicit operator Maybe(null) => new Maybe(/* ... */);
}
public ref struct Maybe
{
    public Maybe(null) => new Maybe(/* ... */);
}
```

For obvious reasons this should not be allowed for classes.

## Motivating Use Case

In [SilkX](https://github.com/Perksey/SilkX) we have [wrapper types representing references/pointers](https://github.com/Perksey/SilkX/blob/367c260e56180e295d049546cb2bd3cd231b4295/sources/Core/Pointers/ConstPtr.cs), that are implicitly convertible from strings and arrays. 

But this isn't the only case, again see Span. The only reason this works today is because of the array operator which feels more of a "happy accident" than a fact known from before the API was designed and/or prototyped.

## Our Dirty Workaround

SilkX defines its own little domain specific language extensions, and `global using static` is used to define a `nullptr` keyword (actually a static property that returns a `NullPtr` that implicitly converts to all of these types. JIT does its thing and inlines all of it).

Usage example:
```cs
var window = Glfw.CreateWindow(1280, 720, "Hello Window!", nullptr, nullptr);
if (window == nullptr)
{
    ConstPtr<sbyte> error = nullptr;
    if (Glfw.GetError(error.AsPtr2D()) == 0 || error == nullptr)
    {
        throw new Exception("failed to create window and failed to get error");
    }

    throw new Exception($"failed to create window: {error}");
}
```

Obviously this is extremely cursed, but this is the only way we could think to achieve the developer experience we're after.

**Comment by 333fred** [comment]
@RikkiGibson has looked at things around this area before. There's also several related issues: https://github.com/dotnet/csharplang/issues/146, https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-07-01.md#non-defaultable-struct-types, https://github.com/dotnet/csharplang/issues/1981, and https://github.com/dotnet/csharplang/issues/2019.

**Reply by HaloFour** [reply]
It's been 6 years but I recall Neal also being opposed to the notion of adding more flavors of `null` to the language: https://github.com/dotnet/csharplang/issues/933#issuecomment-331678693

**Reply by CyrusNajmabadi** [reply]
I'm also personally opposed.  In a very real sense, i appreciate that when operations are done with null (including things like `?.`, `??`, `is null`, `is not null` etc.) i know precisely what semantics that has.

**Reply by RikkiGibson** [reply]
It's just so much work to make the language and runtime respect a user value type's concept of `null`. Including when that type is used as a type argument for a `T` and so on.

I would almost say that we've had several different reasons to want "defaultable value types", "user-defined null on structs", and other highly related scenarios, none of which are compelling enough on their own to do a feature, and which can't all be accommodated by a single design.

**Comment by alahijani** [comment]
Here is one possible design. The literal `null` is always of type `System.Null`, but there is an implicit conversion from `System.Null` to all reference types.

**Comment by TahirAhmadov** [comment]
I'm surprised that `ReadOnlySpan<T>` is implicitly converted to from `null`. I don't think it should. If you want an empty span, either use `ReadOnlySpan<T>.Empty` or `Array.Empty<T>`.

**Reply by CyrusNajmabadi** [reply]
i would certainly expect that it could convert from null.  After all, if i have a `default` instance of ROS, then it's going to have the same bit representation.  So not having this work would be odd.

**Reply by TahirAhmadov** [reply]
To me, it has similar semantics to if something expecting a `string` and when you pass in `null`, it magically converts to `""`. The bit representation is completely irrelevant here - those are implementation details. The high level language semantics are what matter, not the "it's zeroes here and zeroes there, so it should be the same". That's how C++ worked - null, false, 0, etc. are all the same; C# is supposed to be sharper than that. :)

**Reply by TahirAhmadov** [reply]
Also, to the extent that writing `null` is shorter than `ReadOnlySpan<T>.Empty`, the implicit static member syntax (discussed in that other issue, forgot the number) will allow `Empty` in the future; so creating this `null` convertibility only allowed a temporary respite from some typing, at the expense of creating confusion and breaking with other language semantics; not a good trade-off.

**Reply by TahirAhmadov** [reply]
And obviously, with collection expressions, the even shorter `[]` is more clear about what the intention is.

**Reply by RikkiGibson** [reply]
I personally find null converting to ReadOnlySpan via the implicit conversion from array to be basically an unfortunate quirk. I wouldn't want to see people doing that in real code.

**Reply by CyrusNajmabadi** [reply]
> The bit representation is completely irrelevant here - those are implementation details. 

It's not an impl detail.   A default ROS is not an ROS that points at a particular empty array.  They're not the same thing, nor would i expect them to equate to the same thing.  

>To me, it has similar semantics to 

Right.  I agree.  So a ROS wrapping null is not the same as a ROS wrapping an empty array.  That's the same as with ImmutableArray.

> The high level language semantics are what matter

This is literally a low-level feature.  It's intended for advanced users working at a low level.

>  C# is supposed to be sharper than that. :)

C# allows you to use this low level construct *safely*.  That doesn't mean it hides these important details.

**Reply by TahirAhmadov** [reply]
> It's not an impl detail. A default ROS is not an ROS that points at a particular empty array. They're not the same thing, nor would i expect them to equate to the same thing.

> So a ROS wrapping null is not the same as a ROS wrapping an empty array.

How so? From the perspective of code which gets a ROS, it makes no difference whether it's a zero ROS or an empty-array ROS. 
Besides, empty-array ROS aside, something like `ReadOnlySpan<T>.Empty` or just `default` would have been better than `null`.

> This is literally a low-level feature. It's intended for advanced users working at a low level.

It is a high-performance feature in a high-level language. Nothing of value - performance, control, etc. - is gained by allowing to pass in `null` instead of `default`, `ROS<T>.Empty`, or `[]`.

> C# allows you to use this low level construct safely. That doesn't mean it hides these important details.

And here I was this whole time thinking about modularity and abstraction and all that jazz, and now it turns out I need to know the details of what an ROS actually points at. That's just entirely in a different direction from what C# philosophy has been almost the entire time.

**Reply by CyrusNajmabadi** [reply]
> How so? From the perspective of code which gets a ROS, it makes no difference whether it's a zero ROS or an empty-array ROS.

It absolutely makes a difference. These are different spans, and are *observably so*, as part of the documented contract of the API. 

A totally reasonable thing to do, for example, is use the null state to represent a span whose value has not been computed, versus a span that has been computed (but which now points at some empty array).

> It is a high-performance feature in a high-level language. Nothing of value - performance, control, etc. - is gained by allowing to pass in null instead of default, ROS<T>.Empty, or [].

It literally allows for mapping all arrays into the span world, with zero loss of generality (which is a very good thing, and an explicit and intentional part of the API design). 

> , and now it turns out I need to know the details of what an ROS actually points at. 

Yes. You do. It's part of the contract of the type.  You need to understand that contract. 

> That's just entirely in a different direction from what C# philosophy has been almost the entire time.

It absolutely is not. C# and .net *from the beginning* have been about providing performance with safety, entirely by understanding concepts like this. 

You may have felt like you didn't need to take advantage of that things. But that's very different from the language and runtime and apis feeling that they were important details to get right for the rest of the ecosystem to use as effectively as possible.

**Reply by TahirAhmadov** [reply]
> A totally reasonable thing to do, for example, is use the null state to represent a span whose value has not been computed, versus a span that has been computed (but which now points at some empty array).

That's why I said - let's leave `Array.Empty<T>` aside for a second. Why allow to pass in `null`, instead of `ROS<T>.Empty` or `default`?

> It literally allows for mapping all arrays into the span world, with zero lots of generality (which is a very good thing, and an explicit and intentional part of the API design).

I'm not sure I understand. If the implicit operator simply prohibited nullable array values, then passing in `null` would generate an NRT warning (as it should). If you want to pass in `default`, you can do that; and these NRT warnings can prevent bugs if the intention was to pass in a valid array, but somehow a null is passed in. Basically, prohibiting `null` has benefits and doesn't have downsides.

**Reply by TahirAhmadov** [reply]
If you have `int[]? x = ...;` and you want to call `Foo` with either a ROS pointing at `x` if it's not null, or an empty ROS if `x` is null, (an unusual requirement) it would be much clearer to do:
```cs
using System;
public static class C 
{
  static ReadOnlySpan<T> ToROSIfNotNull<T>(this T[]? This)
  {
    if(This != null) return This;
    return default;
  }
  static void Foo(ReadOnlySpan<char> text) { }
  public static void M() 
  {
    char[]? x = null;
    Foo(x.ToROSIfNotNull());
  }
}
```

**Reply by CyrusNajmabadi** [reply]
> That's why I said - let's leave Array.Empty<T> aside for a second. Why allow to pass in null, instead of ROS<T>.Empty or default?

Because if i can't pass in a null instance, then i can't do something as simple as this:

```c#
public void DoSomethingWithArray<T>(T[]? array) => DoSomethingWithSpan(array);

private void DoSomethingWithSpan<T>(ReadOnlySpan<T> span) { ... }

// later:

DoSomethingWithArray(null);
```

Now at every point i want to make a span, i can't just pass null in and have it work.  I have to actually check what values i have and then use `default` or `ROS.Empty`.   As i said above: > It literally allows for mapping all arrays into the span world, with zero lots of generality 

I dont' want to have to special case `null` when moving to the span world.  That would actively be a *bad* thing.  I want to move to the span world *as transparently as possible*, without any sort of loss in generality or any new friction added.

**Reply by CyrusNajmabadi** [reply]
> I'm not sure I understand. If the implicit operator simply prohibited nullable array values, then passing in null would generate an NRT warning

Which would be *awful*.  Because now i can't just operate on all arrays uniformly.  I have to special case them.  Which is an anti goal here.  I should be able to work with all arrays, not a subset of them.  Span isn't the "we got array wrong, especially null array, so here's an alternative that tried to patch that out" feature.  It's the "you can operate on all arrays *and* any sort of contiguous block of memory *uniformly*" feature.  If you break that, you break the feature.

**Reply by CyrusNajmabadi** [reply]
> If you have int[]? x = ...; and you want to call Foo with either a ROS pointing at x if it's not null, or an empty ROS if x is null, (an unusual requirement) it would be much clearer to do:

1. That's much much much less clear. 
2. It requires every point that i'm passing an array to a span to now need this clutter.
3. It adds literal overhead *for no purpose*.

You're starting from the position that null to a span is somehow a bad thing, and attempting to put up unpleasant roadblocks there.  I'm starting from the position that it's uniformly a *good* thing, and that all these efforts are real impediments that actually get in the way of smooth, streamlined, *efficient* coding patterns.

**Reply by TahirAhmadov** [reply]
So basically, this makes it easier to allow accepting spans for existing methods. I understand that goal, but that's a temporary, transitional benefit. Whereas the wart of nulls being convertible to a struct value will stay on forever.

If anything, it should have been designed in a way that a default ROS throws on trying to use it, in which case "converting" null to it maintains the same behavior as before with null-value array references.

> That's much much much less clear.

Have to disagree here. We have a uniform rule in C# that structs cannot become nulls, yet we violate that rule here. At the same time, an appropriately named BCL extension method clearly communicates what's going on.

> It requires every point that i'm passing an array to a span to now need this clutter.

One extension method call? That's a small price to pay for being explicit and stating clearly what's going on.

> It adds literal overhead for no purpose.

Not really, because for example:
```cs
static ReadOnlySpan<T> ToROSIfNotNull<T>(this T[]? This)
{
  return This!;
}
```
This version just wraps around the same implicit operator, and gets optimized away eventually, meaning it's zero overhead.

PS. Also, it looks like too much emphasis on methods expecting a nullable array parameter. I view nullable parameters as more of an exception, not the rule, especially with arrays (or any collections for that matter) - it's a code smell. Most of the time, collections are expected, it should be "not null". Therefore, introducing this null convertibility addresses a scenario which should not be encouraged, which I think is a strange decision.
Oh well....

**Reply by CyrusNajmabadi** [reply]
> We have a uniform rule in C# that structs cannot become nulls,

I don't know what rule you're talking about.

> At the same time, an appropriately named BCL extension method clearly communicates what's going on.

My point is: i don't see the need for an extension, because this is a *normal* and *expected* and *reasonable* state to be in that doesn't require special code.

> This version just wraps around the same implicit operator, and gets optimized away eventually, meaning it's zero overhead.

Why do i need an extension method that literally serves just to remove a NRT warning *that did not need to be tehre in the first place*.  The documented capability of ROS is that is can wrap a null array.  It's *expected* to be able to do that.  We *want* it to be able to do that.  It's a *good* thing that it can do that.  Needing to jump through hoops to make code work that is perfectly sensible and reasonable is just needless friction.

> addresses a scenario which should not be encouraged

I know of nothing that makes me think there is anything about this scenario that should warn or otherwise impede the user.
